main_test.go

```diff
@@ -203,56 +204,188 @@ func TestController(t *testing.T) {
+		Hashring: "hashring0",
+		Tenants:  []string{"foo", "bar"},
+	}}
+	intendedLabels := map[string]string{
+		"manuel": "change",
+	}
+
+	for _, tt := range []struct {
+		name            string
+		hashrings       []receive.HashringConfig
+		labels          map[string]string
+		shouldBeUpdated bool
+	}{
+		{
+			name: "ConfigMapWithHashringChange",
+			hashrings: []receive.HashringConfig{{
+				Hashring: "hashring0",
+				Tenants:  []string{"foo", "bar", "baz"},
+			}},
+			labels:          intendedLabels,
+			shouldBeUpdated: true,
+		},
+		{
+			name:            "ConfigMapWithOtherChange",
+			hashrings:       originalHashrings,
+			labels:          intendedLabels,
+			shouldBeUpdated: false,
+		},
+	} {
+		name := tt.name
+		hashrings := tt.hashrings
+		labels := tt.labels
+		shouldBeUpdated := tt.shouldBeUpdated
+		t.Run(name, func(t *testing.T) {
+			opts := &options{
+				labelKey:               "a",
+				labelValue:             "b",
+				fileName:               "hashrings.json",
+				clusterDomain:          "cluster.local",
+				configMapName:          "original",
+				configMapGeneratedName: "generated",
+				namespace:              "namespace",
+				path:                   "/api/v1/receive",
+				port:                   19291,
+				scheme:                 "http",
 			}
+			klient := fake.NewSimpleClientset()
+			cleanUp := setup(t, klient, opts)
+			defer cleanUp()
+
+			cm := createOriginalConfigMap(t, klient, opts,
+				originalHashrings,
+				[]*appsv1.StatefulSet{
+					{
+						ObjectMeta: metav1.ObjectMeta{
+							Name: "thanos-receive-hashring0",
+							Labels: map[string]string{
+								"a":              "b",
+								hashringLabelKey: "hashring0",
+							},
+						},
+						Spec: appsv1.StatefulSetSpec{
+							Replicas:    intPointer(3),
+							ServiceName: "h0",
+						},
+					},
+				})
 
 			// Reconciliation is async, so we need to wait a bit.
 			<-time.After(500 * time.Millisecond)
-			cm, err = klient.CoreV1().ConfigMaps(opts.namespace).Get(opts.configMapGeneratedName, metav1.GetOptions{})
+			gcm, err := klient.CoreV1().ConfigMaps(opts.namespace).Get(opts.configMapGeneratedName, metav1.GetOptions{})
 			if err != nil {
 				t.Fatalf("got unexpected error getting ConfigMap: %v", err)
 			}
 
-			buf, err = json.Marshal(expected)
+			// Manually change properties of generated configmap.
+			gcm.Labels = labels
+			if _, err = klient.CoreV1().ConfigMaps(opts.namespace).Update(gcm); err != nil {
+				t.Fatalf("got unexpected error updating ConfigMap: %v", err)
+			}
+
+			buf, err := json.Marshal(hashrings)
 			if err != nil {
-				t.Fatalf("got unexpected error marshaling expected hashrings: %v", err)
+				t.Fatalf("got unexpected error marshaling initial hashrings: %v", err)
+			}
+			cm.Data = map[string]string{
+				opts.fileName: string(buf),
+			}
+			if cm, err = klient.CoreV1().ConfigMaps(opts.namespace).Update(cm); err != nil {
+				t.Fatalf("got unexpected error updating ConfigMap: %v", err)
 			}
```

Comment:
This is already done by the `createOriginalConfigMap` func. Why do we do it again?

Reply:
I have put this here because I want to actually trigger a change event without an actual change.

Reply:
If you want to trigger a change, then I think the correct order of operations would be to do:
1. create klient
2. bootstrap initial resources
3. start controller
4. wait for sync
5. compare resources